src/synth/fluid_synth_mono.c

```diff
@@ -0,0 +1,683 @@
+ * used respectively by fluid_preset_noteon() function and voices triggering functions.
+ *
+ * from_key_portamento and from_key_legato are returned by 
+ * fluid_synth_get_fromkey_portamento_legato() function.
+ * fromkey_portamento is set to valid/invalid  key value depending of the portamento
+ * modes (see portamento mode API), CC portamento On/Off, and CC portamento control
+ * (PTC).
+ * Then, depending of the legato modes (see legato mode API), the function will call
+ * the appropriate triggering functions.
+ * @param synth instance.
+ * @param chan MIDI channel number (0 to MIDI channel count - 1).
+ * @param fromkey MIDI note number (0-127).
+ * @param tokey MIDI note number (0-127).
+ * @param vel MIDI velocity (0-127).
+ * @return FLUID_OK on success, FLUID_FAILED otherwise.
+ *
+ * Note: The voices with key 'fromkey' are to be used to play key 'tokey'.
+ * The function is able to play legato through Preset Zone(s) (PZ) and
+ * Instrument Zone(s) (IZ) as far as possible.
+ * When key tokey is outside the current Instrument Zone, Preset Zone, 
+ * current 'fromkey' voices are released. If necessary new voices
+ * are restarted when tokey enters inside new Instrument(s) Zones,Preset Zone(s).
+ * More informations in FluidPolyMono-0004.pdf chapter 4.7 (Appendices).
+ */
+int fluid_synth_noteon_monopoly_legato(fluid_synth_t* synth, int chan,
+							   int fromkey, int tokey, int vel)
+{
+	fluid_channel_t* channel = synth->channel[chan];
+	unsigned char legatomode = channel->legatomode;
+	fluid_voice_t* voice;
+	int i ;
+	/* Gets possible 'fromkey portamento' and possible 'fromkey legato' note  */
+	fromkey = fluid_synth_get_fromkey_portamento_legato( channel, (unsigned char)fromkey);
+
+	if (fluid_channel_is_valid_note(fromkey)) for (i = 0; i < synth->polyphony; i++) 
+	{
+		/* searching fromkey voices: only those who don't have 'note off' */
+		voice = synth->voice[i];
+		if (fluid_voice_is_on(voice) &&
+                    fluid_voice_get_channel(voice) == chan &&
+                    fluid_voice_get_key(voice) == fromkey)
+		{
+			fluid_zone_range_t * zone_range = voice->zone_range;
+			/* Ignores voice when there is no instrument zone (i.e zone_range). Otherwise
+			   checks if tokey is inside the range of the running voice */
+			if (zone_range && fluid_zone_inside_range(zone_range, tokey, vel))
+			{
+				switch (legatomode) 
+				{
+					case FLUID_CHANNEL_LEGATO_MODE_RETRIGGER: /* mode 0 */
+						fluid_voice_release(voice); /* normal release */
+					break;
+					case FLUID_CHANNEL_LEGATO_MODE_MULTI_RETRIGGER: /* mode 1 */
+						/* Skip in attack section */
+						fluid_voice_update_multi_retrigger_attack(voice,tokey,vel);
+					break;
+					default: /* Invalid mode */
+					FLUID_LOG(FLUID_WARN, 
+						"Failed to execute legato mode: %d",legatomode);
+					return FLUID_FAILED;
+				}
+				if (legatomode >= FLUID_CHANNEL_LEGATO_MODE_MULTI_RETRIGGER)	
```

Comment:
I think this if clause should now be moved to the previous switch-case.

Reply:
Right. It will be done.